File: backend/src/main/java/com/atakan/blastscale/common/api/ApiError.java
package com.atakan.blastscale.common.api;

import java.time.Instant;
import java.util.Map;

/**
 * Uniform JSON error body returned by every endpoint.
 *
 * <pre>
 * {
 *   "code": "NO_LIVES_LEFT",
 *   "message": "You have no lives left. Next life in 1240 seconds.",
 *   "details": { "nextLifeInSeconds": 1240 },
 *   "timestamp": "2026-09-04T12:00:00Z",
 *   "path": "/api/v1/levels/42/start"
 * }
 * </pre>
 */
public record ApiError(
        String code,
        String message,
        Map<String, Object> details,
        Instant timestamp,
        String path) {
}

File: unity-client/Assets/Scripts/Net/Dto/ApiError.cs
using System.Collections.Generic;
using Newtonsoft.Json.Linq;

namespace BlastScale.Client.Net.Dto
{
    /// <summary>
    /// Uniform JSON error body returned by every endpoint (mirrors <c>ApiError.java</c>).
    /// DTO field names are deliberately camelCase: they are the JSON contract, one-to-one.
    /// </summary>
    public sealed class ApiError
    {
        public string code;
        public string message;
        public Dictionary<string, JToken> details;
        public string timestamp;
        public string path;
    }
}
